### ISSUE TITLE
no support for temperature / thermostat

### ISSUE THREAD

**Issue by JurikRistjouw** [giejay/domoticz-gbridge-plugin#3]
I've added a temperature sensors and a thermostat device, but i get these errors from domoticz log:

2018-12-14 11:16:50.930 Error: (kapelt gBridge) No gBridge adapter found for device: Room Temperature
2018-12-14 11:16:50.930 Error: (kapelt gBridge) No gBridge adapter found for device: Living Room Thermostat

according to the docs, it should work. Is this a bug or not yet implemented?

kind regards,

Jurik

**Comment by giejay** [comment]
Woops, I have a temp adapter implemented but I forgot to add the adapter to the list. Will do that tonight.

**Comment by JurikRistjouw** [comment]
nice! can't wait to pull ;)

**Comment by giejay** [comment]
I added support for Setpoints and Im able to set the temperature of my room. Asking what the temperature currently is, is not supported yet.

**Comment by JurikRistjouw** [comment]
i keep having some trouble, what command are you giving to make the setpoint change?

**Comment by giejay** [comment]
Hey Google, set the 'device name' to 22 works for me

**Comment by JurikRistjouw** [comment]
Ah, thanx

mine is called: "Living Room Thermostat". When I call for set living room thermostat to 22. I get: this mode isn't available for 2 thermostats (i only have one, i've also opened an issue for that on gbridge github)

edit: after deleting the device from the kapelt device page and update on the domoticz hardware tab, i get a response like this: "this mode isn't available for the living room thermostat"

**Comment by giejay** [comment]
I had the same issue, what worked for me was setting a nickname: room and then saying set the room to 22

**Comment by JurikRistjouw** [comment]
I am trying to rename the device, but getting some trouble deleting the old device from device list in gbridge and refreshing. Will keep trying

**Comment by giejay** [comment]
Ah, I was not very clear, I renamed the device in google home, under devices. You can set a nickname over there

**Comment by JurikRistjouw** [comment]
ah right, thanx! I will try that.

**Comment by JurikRistjouw** [comment]
I am still having the same issue, even after nicknaming in google. Mode is not supported on the "nickname". (the other stuff works brilliant! Thanx)

**Comment by giejay** [comment]
When double checking, I saw that only the setpoint trait was added by the plugin. I fixed that with the latest commit and Im now able to set my temperature without any nickname tricks. Can you try again by pulling and clicking update in the hardware section? Make sure to remove the device first from gBridge device page or give it a different name in Domoticz (description)

**Comment by JurikRistjouw** [comment]
I had to delete the device 2 times (maybe google cache), but after that pulled and "clicked update on domoticz". Device appeared in gBrigde with more traits. mode, setpoint and ambient. It works now without a nickname. I can even say set the room to x degrees, and it "finds" the thermostat.

**Comment by JurikRistjouw** [comment]
hmm Saying: "lower the living room temperature by 1 degree" -> "that mode isn't suported for the "device name".... but maybe that is a feature request ;) 

edit: when saying: what is the current temperature in the room, or device, I get: unable to reach device name...

**Comment by giejay** [comment]
For the first one: you have to say: lower/increase the temperature in the 'devicename' by 1 (degrees). Have no idea why google didn't make the other one work. Then you will see that Domoticz tries to set the temperature to minus 1/plus 1, thats where the non implemented part comes:) Thankfully I could make a relative small change to also get that working, I'm testing that now on my test environment and will push it this afternoon.

The second one, I'm not sure about. That has no direct relation to this plugin so that seems to be a gBridge issue/Google issue. I'm also not able to get the temperature in the room... 

Maybe @peterkappelt can help us?

**Comment by giejay** [comment]
Just pushed the change, can you try again?

**Comment by JurikRistjouw** [comment]
I will try again.

**Comment by JurikRistjouw** [comment]
no luck as of yet. that mode isn't supported on "device"

**Comment by giejay** [comment]
Very weird, this is my flow:

- increase the temperature in the room by 1

- You got it, warming up the room.

No idea why that wouldn't work for you

**Comment by JurikRistjouw** [comment]
i am not sure if it is a cache issue with google or gbridge. Somehow i am unable to properly delete a device (the thermostat) from gbridge to re-add it. Although it says delete succesfull, after a fresh of the page, the device is back. Not sure if this is causing the issue. I restarted domoticz after a pull to make sure. and synced the devices twice.

**Comment by giejay** [comment]
I had the same issue when readding a device, it kept quering domoticz with the old id so domoticz couldn't find it. After a couple of hours it started sending the new device id so I think it's a cache issue on google side

**Comment by JurikRistjouw** [comment]
ok, let's see how it behaves tomorrow.

**Comment by JurikRistjouw** [comment]
still the same :-( i also see no ambient temperature (which makes sense in a way, because it is only a setpoint, but I also don't see the setpoint returned)) Not sure how to assist.

**Comment by peterkappelt** [comment]
I'd be happy to help.

> 
> 
> i am not sure if it is a cache issue with google or gbridge. Somehow i am unable to properly delete a device (the thermostat) from gbridge to re-add it. Although it says delete succesfull, after a fresh of the page, the device is back. Not sure if this is causing the issue. I restarted domoticz after a pull to make sure. and synced the devices twice.

Where are you unable to delete a device? In the gBridge dashboard, in the Google home app or with the Domoticz plugin?



The ambient temperature is never shown in the Google home app, you can only get it by querying. It is "Hey Google, what's the temperature in here?" or "Hey Google, wie warm ist hier?" (in German).

For the issue with devices not syncing properly: There is a command: "Hey Google, synchronisiere meine Geräte"/ "Hey Google, synchronize my devices". It seemed to accelerate the Google syncing a bit.



> edit: when saying: what is the current temperature in the room, or device, I get: unable to reach device name...

Could you please give me your gBridge user id and try to reproduce this issue with creating the device in the gBridge dashboard and manually sending the MQTT messages for it?

**Comment by JurikRistjouw** [comment]
I was unable to delete the device from the gbridge device page. It says delete succesfull, but it keeps returning. But, when deleted, it reappears as a new device in google. Then I readd it to my room.

I can understand that, the ambient temp is not displayed in the setpoint device. Only the setpoint could be read, but that is not the case. In my case a remeha iSense controlled through open therm gateway, the device is only setpoint, not ambient also. There is a sensor for the ambient, but not sure how to match them up so google and or gbridge will understand.

I have tried "sync my devices" twice... a few times now. Also pushed from the hardware tab in domoticz, restarted domoticz and waited. Results are the same. 

my id is uniform one six three (hope you get that)

Have tried to reproduce the issue just after posting, not sure how to manually send an mqtt message tho...

**Comment by peterkappelt** [comment]
The issue with devices reappearing seems to be caused from this plugin. I've checked the logs, a new device seemed to be added via the API. I'd have to look into this plugin, I think I can do it next week.

The problem that Google says "unable to reach device" should actually never happen. However, I'm not able to find anything suspicious in the logs. I can't reproduce it for myself, so debugging is hard.

The process for manually testing would look like that:
* Disable the domoticz gBridge plugin
* Download an MQTT client, e.g. MQTT.fx
* Connect to the gBridge MQTT server with it, using the data displayed in "My Account" in the portal
* Manually create a thermostat in the portal
* Send messages on the matching topics, containing temperature data or alike. Query Google and check, whether your sent data is applied and the device is still marked as reachable.

@giejay Could you reproduce the problem that the device is not reachable anymore?

**Comment by giejay** [comment]
I am able to reproduce the issue regarding, what's the temperature in here. Google responds with: Im not sure which device you would like to control.

Regarding adding of devices, its true that after you say: sync my devices, this plugin will add devices from domoticz to gBridge, so that probably explains why the device can't be removed. Disable the plugin in the hardware as Peter mentioned.

**Comment by JurikRistjouw** [comment]
I will try to remove the device as mentioned with the plugin disabled and readd it. 

@peterkappelt It is able to reach the device, but it when asking the current temp it responds with: that mode isn't supported for "device name" or device isn't available right now. When raising the temp it responds with: that mode isn't supported on "device name" When setting the thermostat, it responds with: sure, setting the "device name" to xx.x degrees. 

I will mention again, that only one thermostat is added to gbridge, not an ambient temperature sensor. Hence it cannot display or return the current temperature, only the setpoint. I could add the ambient sensor and give it the same name?? would that work?

**Comment by JurikRistjouw** [comment]
Seems that a temp sensor isn't supported at this time. This was in the logs

`2018-12-21 18:08:49.475 Error: (kapelt gBridge) No gBridge adapter found for device: {'Description': 'gBridge:Living Room Thermostat', 'Data': '19.2 C', 'XOffset': '0', 'PlanIDs': [4, 10], 'HardwareID': 8, 'BatteryLevel': 255, 'AddjMulti': 1.0, 'HardwareTypeVal': 20, 'HaveTimeout': False, 'Name': 'otgw Room Temperature', 'YOffset': '0', 'Temp': 19.2, 'Timers': 'false', 'SubType': 'LaCrosse TX3', 'HardwareType': 'OpenTherm Gateway with LAN interface', 'trend': 1, 'idx': '105', 'AddjMulti2': 1.0, 'HardwareName': 'OpenTherm', 'SignalLevel': '-', 'Unit': 8, 'Type': 'Temp', 'Used': 1, 'AddjValue2': 0.0, 'PlanID': '4', 'Notifications': 'false', 'LastUpdate': '2018-12-21 18:08:28', 'TypeImg': 'temperature', 'ShowNotifications': True, 'CustomImage': 0, 'Protected': False, 'Favorite': 1, 'AddjValue': 0.0, 'ID': '0008'}`

WARNING: when giving the same description for both sensor and thermostat; the second device isn't pushed anymore.. I could not find the thermostat in gbridge. I removed the same description from the sensor which wasn't supported, and it came back instantly in gbridge. 

will test further

**Comment by giejay** [comment]
Correct, only SetPoint devices are supported for now.

**Comment by JurikRistjouw** [comment]
ok, clear.

I've come to allmost the same conclusion. Only one thermostat is available and visible in gbridge and google home.

When asking the temperature i get a different response: I am sorry, I am not sure which device you want to control

cannot raise or lower the setpoint, the response is: That mode isn't available for two thermostats.

I can set the temperature, but the response is: Sure, setting the "device name" to xx.x degrees. That mode isn't available for "device"

**Comment by giejay** [comment]
Can you try readding the device under a different name so we are sure that it's not a google cache issue. What is your GH language? English or German?

**Comment by JurikRistjouw** [comment]
I will try relinking google regarding the other issue. i guess that should take care of the cache issue? I will add the the thermostat with a different name (currently called living room thermostat). My GH language is English and Dutch.

**Comment by JurikRistjouw** [comment]
after relinking and adding devices. 

asking: lower the temperature on x-> that mode isn't supported for the x
asking: set the x temp to xx.x degrees -> you got it, setting the x to xx.x degrees
asking: what is the temperature in the x -> I am sorry, I am not sure which device you want me to control

hope this helps

**Comment by giejay** [comment]
Some development at my side. My temperature setpoint device is called 'Room'. Previously when I asked, whats the temperature in the room, it returned with:  'I am sorry, I am not sure which device you want me to control', just like you. I noticed I had two thermostats in my GH devices (one was a test from my dev machine), I removed that one and now I'm getting: The room is off and its currently zero degrees.

When I go the thermostat in the Google Home app, it states that the ambient temperature is 0 and the device is off, so that seems to make some sense. @peterkappelt Should the plugin do something with query requests (next to commands)? All the data is cached in redis right? So when I say, set the temperature to 22, it saves that in the Redis DB and that cached value is returned when Google queries gBridge, right?

**Comment by JurikRistjouw** [comment]
I still have the issue with: i am not sure which device you like me to control, afaik only one thermostat is in gbrigde.

when you use cached values, will it pick up a manual setpoint alteration (by hand)? Domoticz will pick that up, but is that also updated to Redis DB from gbridge?

**Comment by giejay** [comment]
No, thats something that still needs to be implemented, I just don't know how to listen to device changes from my plugin, the only thing I can do is just publish the current state every x minutes/seconds but except for the temperate, which does not work anyway, I don't see the need yet.

**Comment by giejay** [comment]
I just tested it with a light, asking whether it was on, right after I asked google to turn it on. Unfortunately, it says the device is off so it doesn't seem to be only thermostats which dont work

**Comment by JurikRistjouw** [comment]
I have the same conclusion, asking if the lights are on in room x it names the 2 lights only (good) and tells me they are off. Right after that it says, unable to reach 6 lights.... (i guess cache issue for the numbers)

**Comment by JurikRistjouw** [comment]
update: it is definitly cache related somewhere. When i ask now: "set the temperature in the living room to x degrees" it responds: "sure, setting the living heating (new name) to x degrees" "Unable to reach living room thermostat (previous name" 

also, the count in the attic is trippled somehow...

please advice how to solve

**Comment by giejay** [comment]
I'm afraid I can't really help you regarding cache issues, I have the same problem, if I say turn off the outside shutter, it says, sure, turning off 2 things. I guess we will just have to wait it out or something

**Comment by JurikRistjouw** [comment]
isn't there any way to reset google's cache? maybe @peterkappelt knows a trick?

**Comment by JurikRistjouw** [comment]
https://support.google.com/googlehome/answer/7073477?hl=en

maybe that could help?

**Comment by peterkappelt** [comment]
Well, I've got some more time now, since the subscription project is done. Let me tell you, providing and preparing SaaS-products is harder than it seems at a first glance...

However, I'm going to take a day off (tomorrow) and then I'll have a deeper look into this issue, trying to reproduce everything. I expect it to be a combination of a small bug in gBridge, general issues with Google Assistant and a possible misunderstanding about the functionality.

Merry christmas everyone!

**Comment by JurikRistjouw** [comment]
I can imagine SaaS is hard. Production is something else compared to hobby. Thank you for taking the time to dig into it deeper. I couldn't agree more with your conclusion!

Merry Christmas!

**Comment by peterkappelt** [comment]
> I can imagine SaaS is hard. Production is something else compared to hobby.

Yeah... Actually I'm doing electrical engineering for a living - that is somehow completely different.

I did some experiments this evening. Apparently Google uses different APIs for smart home providers being available to the public (hosted service) and private ones (my development server/ self hosted setups), hence I couldn't reproduce most of your issues on my development system. I tried it on the hosted product  - et voilà: I could see the most problems.

As it seems, my assumption was right. There was a small bug in a gBridge component, I fixed it just now. Possibly, the Google caching issue doesn't appear anymore with the gBridge problem bein solved. 

And there is a misconception ;). I'm going to write about it tomorrow...

**Comment by giejay** [comment]
I'm very curious what the fix could be, where did you apply it? I see the latest commit is 4 days ago so not in code?

**Comment by peterkappelt** [comment]
Well, first about the misconception.

Once you say an Execute-Command to Google (Set x to ..., Turn z on/ off, ...), this value just gets published to the matching MQTT action topic (topics *without* set appended). Nothing is set in the redis cache for that case.

When you publish something to an MQTT status topic (topics with set appended), this value is indeed stored in the redis cache. Once you say a query-command (Is x turned on?), this cached data gets transfered to Google.

These are two completely separated data paths. In my opinion, in a theoretical and abstract point of view, this is the right way to do for any event driven smart home software. When you ask Google to turn a device on, that doesn't necessarily mean that the device gets turned on - it could be offline, blocked by a fixed timing schedule, in some sort of manual control mode, etc.
When you use the described way, you just need to follow a plain structure. If data is received on an action topic, (try) to set the device to that. Everytime a parameter/ setting of your device changes, report that change to the matching status topic, no matter where that change originated from.

When no data is cached (via publishing to the status topics), gBridge uses default values: mode is "off", temperature setpoint and ambient temperature are both "0.0". However, Google reacts strangely to questions about the thermostat if not all parameters are set different from the default values. 

Long story short: In order for this to work properly, you need to report some data back to the status topic. The prefered way would be to set up some sort of event listener for the controlled devices, that sends the current data to this topic once something changes for this device.

The easiest way would be to just send back the data to status topic, right at the moment when you receive something on the action topic. Just send some data to the status topic of mode, ambient temperature and setpoint. I expect it to work then.

**Comment by peterkappelt** [comment]
So, now for the bug of gBridge:
I’ve written that once data is sent on the status topic, it gets cached and sent to Google once a query is said.
However, if publishing your app publicly, Google requires you to implement the “Report State” feature aside from that (https://developers.google.com/actions/smarthome/report-state). When we receive something on the status topic, we also need to notify Google proactively about the new device states for that user (besides from storing it in the cache for further queries).
This part of gBridge is not publicly available, since it is not required for self-hosted setups. In self-hosted setups, Google just queries the cache of the gBridge instance. In the public version, Google wants to keep its own cache.
The problem was in this part, it happened under certain load conditions. When two different users sent data to a status topic at the same time (in combination with high system load), it was possible that their data was mixed up. Sometimes, the device states for user A that were sent to Google contained device states from user B.  This is no security breach, since the Google API always rejected those invalid records – but with an undocumented error code, that wasn’t catched by the gBridge app. 
The problem of mixing data were two asynchronous node.js functions, influencing each other. I fixed the problem by replacing promise-/ callback-based functions with the newer await/ async constructs; while adding more code to catch unexpected error codes.

**Comment by giejay** [comment]
Thanks for the very detailed information! I will try to add support for pushing changes also to the status topics, once a device is (successfully) updated through gBridge, and also when it's updated though Domoticz (app).

**Comment by JurikRistjouw** [comment]
Device count is correct now and no more double thermostat. Thanx for fixing and explaining!

**Comment by giejay** [comment]
With the 2 latest commits, you can now ask what the temperature in the room is, and also if a light/switch is on :) 

It still says that the room is off (something to do with the heat mode?) but it does tell you the temperature.

**Comment by TheReptile** [comment]
The latest update breaks gBridge.
On startup of Domoticz:
```
2018-12-27 21:00:19.285 Error: (gBridge) 'onStart' failed 'JSONDecodeError'.
2018-12-27 21:00:19.285 Error: (gBridge) ----> Line 157 in '/home/pi/domoticz/plugins/gbridge/plugin.py', function onStart
2018-12-27 21:00:19.285 Error: (gBridge) ----> Line 71 in '/home/pi/domoticz/plugins/gbridge/plugin.py', function onStart
2018-12-27 21:00:19.285 Error: (gBridge) ----> Line 77 in '/home/pi/domoticz/plugins/gbridge/plugin.py', function syncDevices
2018-12-27 21:00:19.285 Error: (gBridge) ----> Line 48 in '/home/pi/domoticz/plugins/gbridge/gbridge_client.py', function fetchDevicesFromBridge
2018-12-27 21:00:19.285 Error: (gBridge) ----> Line 319 in '/usr/lib/python3.5/json/__init__.py', function loads
2018-12-27 21:00:19.285 Error: (gBridge) ----> Line 339 in '/usr/lib/python3.5/json/decoder.py', function decode
2018-12-27 21:00:19.285 Error: (gBridge) ----> Line 357 in '/usr/lib/python3.5/json/decoder.py', function raw_decode 
```
When using a voice command:
```

2018-12-27 21:04:12.803 Error: (gBridge) 'onMessage' failed 'TypeError':'argument of type 'NoneType' is not iterable'.
2018-12-27 21:04:12.803 Error: (gBridge) ----> Line 178 in '/home/pi/domoticz/plugins/gbridge/plugin.py', function onMessage
2018-12-27 21:04:12.803 Error: (gBridge) ----> Line 100 in '/home/pi/domoticz/plugins/gbridge/plugin.py', function onMessage
2018-12-27 21:04:12.803 Error: (gBridge) ----> Line 126 in '/home/pi/domoticz/plugins/gbridge/mqtt.py', function onMessage
2018-12-27 21:04:12.803 Error: (gBridge) ----> Line 135 in '/home/pi/domoticz/plugins/gbridge/plugin.py', function onMQTTPublish
2018-12-27 21:04:12.854 Error: (gBridge) 'onMessage' failed 'TypeError':'argument of type 'NoneType' is not iterable'.
2018-12-27 21:04:12.854 Error: (gBridge) ----> Line 178 in '/home/pi/domoticz/plugins/gbridge/plugin.py', function onMessage
2018-12-27 21:04:12.854 Error: (gBridge) ----> Line 100 in '/home/pi/domoticz/plugins/gbridge/plugin.py', function onMessage
2018-12-27 21:04:12.854 Error: (gBridge) ----> Line 126 in '/home/pi/domoticz/plugins/gbridge/mqtt.py', function onMessage
2018-12-27 21:04:12.854 Error: (gBridge) ----> Line 135 in '/home/pi/domoticz/plugins/gbridge/plugin.py', function onMQTTPublish
2018-12-27 21:04:14.003 Status: MySensors: Using serial port: /dev/ttyUSB1
```

**Comment by JurikRistjouw** [comment]
> With the 2 latest commits, you can now ask what the temperature in the room is, and also if a light/switch is on :)
> 
> It still says that the room is off (something to do with the heat mode?) but it does tell you the temperature.

That sounds awesome! But I think you mean setpoint there? So I can ask what the current setpoint is.. temperature was not supported and I therefor have no temp sensors added. Could you confirm the code is breaking at this point?

**Comment by JurikRistjouw** [comment]
@giejay would it be possible to have a development branch in git? so we can switch back to master if it breaks in the future?

**Comment by giejay** [comment]
@JurikRistjouw  Yes I can make a develop branch.

@TheReptile For me everything works, on my local setup and the hosted gBridge. I've added debug logging in the latest commit, can you pull, update and show me the url which is printed: 'Calling API: http://www'. Off course, enable debug logging;)

**Comment by JurikRistjouw** [comment]
@giejay awesome! I will try the latest commit. looking forward to the development branch

**Comment by TheReptile** [comment]
@giejay Ok, seems I found the configuration error. In my settings I had https://gbridge.kappelt.net/ (with trailing /). That broke the communication. After I removed the / everything worked. Something must have changed, since this worked before.

**Comment by JurikRistjouw** [comment]
ok, pulled latest commit (include action in debug logging). Nothing breaks, but after push and sync twice, I get unable to reach thermostat or devices are off (when they are on)

**Comment by giejay** [comment]
When do you get that response? What do you ask?

**Comment by JurikRistjouw** [comment]
when asking: what is the temperature in the living room: unable to reach the "device name"
when asking what lights are on in the attic: the x and y are off.  (should be on)

EDIT, the thermostat was added as device to the living room in GH
When asking, set the living room temp to 20.5 degrees, it responds: you got it, settting the "device name" to 20.5 degrees

**Comment by giejay** [comment]
Did you first turn on the temperature and the lights through Google? That's needed for now, until I built in a timer to update statuses in gBridge/google every x minutes.

**Comment by JurikRistjouw** [comment]
I did not set a temp when rebooted. I guess i've done it now because it works. Now i can raise and lower the temp as well as get a setpoint back. I say close the issue for now, or does it need work to respond on the manual overrides?

edit
btw, it has some cache issues still. i've set the setpoint to 18. asked to raise by 0.1 (setpoint is 18.1 in domoticz now) when asking what is the current temp. it responds 18. it should have responded 18.1.

**Comment by giejay** [comment]
that might be a rounding issue, do you get the same behavior as setting 18 to 19?

**Comment by JurikRistjouw** [comment]
check, when setting a full degree by voice it responds and operates correctly. only when settings by hand it does not transfer to gbridge and thus the temp is at the previous (gbridge aware) setpoint. As you said it is at this point. Thus far, it is as good is at is can be i guess. cheers!

**Comment by JurikRistjouw** [comment]
btw, i switched to the develop branch! thanx

**Comment by giejay** [comment]
Thanks to the PR of Sylvain i could integrate support for pushing back temperature changes to the "set" topic, also if the temp wasn't changed through gBridge but through Domoticz for example. So this should now fully work, can you test?

**Comment by JurikRistjouw** [comment]
@giejay I will pull, test and let you know the results.

**Comment by JurikRistjouw** [comment]
well domoticz crashed on me after resetting the branch and pulling the latest commit. Any advice? should i clear the __pycache directory or do something else?

up and running again after a reboot... keep you posted

**Comment by JurikRistjouw** [comment]
It seems the temp can be set with voice, but is not returned when asked. It looks like it reads out the current temperature, instead of the setpoint. I don't know how to differentiate. I tried, setpoint and thermostat with no luck. Only temperature can be asked. Also not updated when set with domoticz. But maybe that is due to the fact, it reads out the current temp instead of the setpoint...

**Comment by giejay** [comment]
Did you add the mqtt plugin to domoticz hardware?

**Comment by JurikRistjouw** [comment]
well, it was still there? should i readd it?

**Comment by giejay** [comment]
No I mean the MQTT Client Gateway with LAN interface hardware. This is now used for returning the values to gBridge.

**Comment by JurikRistjouw** [comment]
I am not sure what you mean. I have the kapelt gbridge account with mqtt from the beginning. Should i add another?

**Comment by giejay** [comment]
Go to the hardware page and add the MQTT Client Gateway hardware using the same mqtt host and port as you are using gBridge

**Comment by JurikRistjouw** [comment]
not sure what you want me to do, I already have the kapelt gbridge in domoticz. Do i have to add another account or just refresh the current mqtt plugin from domoticz by disabling and re-enabling the device?

anyway I get errors, see next post

**Comment by JurikRistjouw** [comment]
i tried adding the same credentials for a generic mqqt plugin, but i get errors

**Comment by JurikRistjouw** [comment]
2019-04-07 00:08:30.194 Error: (kapelt gBridge) 'onMessage' failed 'TypeError':''int' object is not subscriptable'.
2019-04-07 00:08:30.194 Error: (kapelt gBridge) ----> Line 198 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/plugin.py', function onMessage
2019-04-07 00:08:30.194 Error: (kapelt gBridge) ----> Line 105 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/plugin.py', function onMessage
2019-04-07 00:08:30.194 Error: (kapelt gBridge) ----> Line 126 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/mqtt.py', function onMessage
2019-04-07 00:08:30.194 Error: (kapelt gBridge) ----> Line 166 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/plugin.py', function onMQTTPublish
2019-04-07 00:08:30.194 Error: (kapelt gBridge) ----> Line 37 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/adapters/on_off_switch_adapter.py', function publishState
2019-04-07 00:08:30.216 Error: (kapelt gBridge) 'onMessage' failed 'TypeError':''int' object is not subscriptable'.
2019-04-07 00:08:30.216 Error: (kapelt gBridge) ----> Line 198 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/plugin.py', function onMessage
2019-04-07 00:08:30.216 Error: (kapelt gBridge) ----> Line 105 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/plugin.py', function onMessage
2019-04-07 00:08:30.217 Error: (kapelt gBridge) ----> Line 126 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/mqtt.py', function onMessage
2019-04-07 00:08:30.217 Error: (kapelt gBridge) ----> Line 166 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/plugin.py', function onMQTTPublish
2019-04-07 00:08:30.217 Error: (kapelt gBridge) ----> Line 37 in '/home/pi/domoticz/plugins/domoticz-gbridge-plugin/adapters/on_off_switch_adapter.py', function publishState

**Comment by giejay** [comment]
Whats the incoming message thats appearing in the debug line just above the error?

**Comment by giejay** [comment]
I have updated the readme, can you check if you executed step 6 correctly?

**Comment by JurikRistjouw** [comment]
> Whats the incoming message thats appearing in the debug line just above the error?

not related to kapelt gbridge, differs with reboots. This time it is a failed login attempt to solaredge.

**Comment by JurikRistjouw** [comment]
> I have updated the readme, can you check if you executed step 6 correctly?

will do

**Comment by JurikRistjouw** [comment]
@giejay I don't use a local mqtt server. I use the one from kapelt. Not sure how to do step 6

**Comment by giejay** [comment]
Ah sorry, should have started with that:) 

Maybe if we ask @peterkappelt nicely, he can reserve some space (with ACL) for "domoticz/out/{userid}" in his MQTT broker :) Then it would also work for the remote gBridge.

For now, I discovered a bug when not using the local domoticz topic so that might explain why its not working for you. I have fixed the bug and will test it tonight when Im home.

**Comment by SylvainPer** [comment]
If you can host a mqtt server, this can be a solution: https://doc.gbridge.io/firstSteps/mqttServer.html#using-own-mosquitto-as-a-bridge

**Comment by giejay** [comment]
If we could add extra parameters to the plugin (which is not possible due to the limit of 6), we could add an extra local MQTT server for separating the domoticz/out vs the gbridge topic. Maybe a PR on Domoticz to make the limit go away:)

**Comment by SylvainPer** [comment]
Yes, it will avoid the test on the server name that is not the better way...

**Comment by SylvainPer** [comment]
Another way could be two branches on github, not the best way but it'll work.

**Comment by JurikRistjouw** [comment]
I have someone magically install the "proxy" mqtt (thanx for that the one who's name cannot be mentioned) and it works!! it reads out the status correctly now. Being really critical, I would be nice to see it differentiate when asking the set-point or temperature. Now it reacts to set temperature, and it will set the set-point. But i guess that is an issue with gbridge itself?

@giejay fantastic work! thanx

**Comment by giejay** [comment]
No thats not an issue with gBridge, thats an issue in this code:
```
 if 'svalue1' in message:
            temp = self.get_temperature(message['svalue1'], device)
            topic = topic + '/' + str(message['idx'])
            mqtt_client.Publish(topic + '/tempset-ambient/set', temp)
            mqtt_client.Publish(topic + '/tempset-setpoint/set', temp)

```
I will fix this as well by checking the message for a type (hopefully) like Setpoint vs temperature.

**Comment by JurikRistjouw** [comment]
@giejay Awesome! This thread keeps getting better and better! Waiting patiently

**Comment by giejay** [comment]
I pushed my changes to develop.

One thing I noticed that my Temperature + Setpoint devices are split up for my Tado Thermostat, so I should actually also mark the Temp device as a "gBridge" device and probably give it a slightly different name to be able to query the room temperature.

**Comment by JurikRistjouw** [comment]
pulled and rebooted, have the same issue as you. I added the temp sensor of the thermostat also. Both have a setpoint now... not sure how to name them. will test further.

**Comment by SylvainPer** [comment]
The problem is as they are split, you create 2 devices.
I think there are several possibilities:
 - create a new keyword "gBridge thermostat: x", with x the ambient idx
 - create a config file (not user friendly)
 - modify the gbridge config to set the ambient with the good idx but you'll create an unnecessary device
 Another idea ?

**Comment by JurikRistjouw** [comment]
i concur that it is impossible for gbridge/plugin to know which is which. And no way in google to differentiate. Both are registered as temp devices.

**Comment by SylvainPer** [comment]
Google is not the problem, there is a way to ask: https://doc.gbridge.io/traits/temperaturesetting.html
The plugin is not yet able to manage it but as it's software, we have to find a solution.

I don't have a thermostat for the moment, I've a dedicated arduino that works perfectly so I don't know how many devices you have. Three ? one for the mode, one for the ambient and one for the setpoint ?

**Comment by giejay** [comment]
I have introduced a feature called "LinkedDevices"

Add linkedDevices:{indexOfDevice} to the description, for example: linkedDevices:13 to the device with index 12 (the thermostat). The index 13 is the Temperature device. This results in:

2019-04-06 23:56:31.477 (gBridge) MQTT message: domoticz/out {'svalue3': '1', 'idx': 13, 'stype': 'WTGR800', 'id': '3', 'nvalue': 0, 'svalue2': '50', 'svalue1': '22.1', 'description': '', 'dtype': 'Temp + Humidity', 'RSSI': 12, 'name': 'Thuis Woonkamer TempHum', 'unit': 0, 'Battery': 255}
2019-04-06 21:56:31.478 (gBridge) MqttClient::Publish gBridge/u1/12/tempset-ambient/set (22.1)
2019-04-06 21:56:31.479 (gBridge) MqttClient::Publish gBridge/u1/12/tempset-humidity/set (50)

Fixed:)

Btw, you can add multiple linked devices, so for device 11: linkedDevices:12,17,20.

All messages for 12,17 and 20 will be handled by the handler for the device 11

**Comment by JurikRistjouw** [comment]
pfoe, another plugin... sounds like a meme ;) at this point i have the temp set device and temp read device from the thermostat added to gbridge.

![image](https://user-images.githubusercontent.com/8580506/55757345-00484600-5a54-11e9-9e35-e258d37da72a.png)


EDIT: deleted the second device after @giejay comment and testing

**Comment by giejay** [comment]
Another plugin? What do you mean? You dont need another plugin for this feature

**Comment by JurikRistjouw** [comment]
> linkedDevices:13

Will try that feature! added the ID of the temp senor to the thermostat on a new line after the friendly name.

**Comment by JurikRistjouw** [comment]
is the second id (the senor) supposed to be exposed for gbridge? and if so, what name

I can set the temp, but when asking the temperature it states that the living heating is unreachable...

how do you ask for setpoint versus reading out the temp?

**Comment by giejay** [comment]
No not necessary

**Comment by giejay** [comment]
```
Query﻿ ﻿thermostat:
what﻿ is <﻿device> set to
what is <device>﻿ inside temperature
what is <device> ﻿humidity (If your thermostat supports it)
 
Change mode:
Set <device> mode to <off|heat|cool|auto|on>
Turn off <device>
Turn on <device>
 ﻿
Change temperature:﻿
Set <device> to <temp>
Increase <device> by <tempDelta﻿>
 
Multiple thermostat commands:﻿
What is the inside temperature
Set heating to <temp>﻿
Set cooling to <temp>﻿﻿
Set the thermostat modes to <mode﻿>﻿﻿﻿﻿﻿
```

**Comment by JurikRistjouw** [comment]
hmm cannot reach livinig heating when asking to report the temperature. setting works fine

**Comment by giejay** [comment]
When you open it in the Google home app, what do you see when checking the Living Heating?

**Comment by JurikRistjouw** [comment]
I get this.. 
![photo_2019-04-09_23-42-52](https://user-images.githubusercontent.com/8580506/55837282-356e9a00-5b21-11e9-9257-a0c78a488476.jpg)

this is the living room thermostat, with the added temp senor through friendly name like so.
gBridge:Living Heating 
linkedDevices:105

when asking the temperature of the living room, i get: not sure which device you want me to control

**Comment by giejay** [comment]
Are you sure you are seeing this message: 
2019-04-06 21:56:31.478 (gBridge) MqttClient::Publish gBridge/u1/12/tempset-ambient/set (22.1)
Whenever a change for device 105 occurs?

Did you also restart the plugin by clicking update in the hardware tab? Or just rebooting domoticz

**Comment by JurikRistjouw** [comment]
I am not sure, will try to adapt and test some more.. I did pull the latest commits, updated/synced in gbridge and rebooted ;-)
